File: WPF/Windows SDK WPF Samples/AppModel/FixedHierarchicalTopologySample/VisualBasic/TaskLauncher.vb
Imports System
Imports System.Windows.Navigation

Public Class TaskLauncher
    Inherits PageFunction(Of TaskContext)

    Public Sub New()
        Me.taskData = New TaskData
    End Sub

    Protected Overrides Sub Start()
        MyBase.Start()
        MyBase.KeepAlive = True
        Dim page1 As New TaskPage1(Me.taskData)
        AddHandler page1.Return, New ReturnEventHandler(Of TaskResult)(AddressOf Me.taskPage_Return)
        MyBase.NavigationService.Navigate(page1)
    End Sub

    Public Sub taskPage_Return(ByVal sender As Object, ByVal e As ReturnEventArgs(Of TaskResult))
        Me.OnReturn(New ReturnEventArgs(Of TaskContext)(New TaskContext(e.Result, Me.taskData)))
    End Sub

    Private taskData As TaskData

End Class

File: WPF/Windows SDK WPF Samples/AppModel/FixedLinearTopologySample/CSharp/TaskLauncher.cs
namespace FixedLinearTopologySample
{
    using System;
    using System.Windows.Navigation;
    
    public class TaskLauncher : PageFunction<TaskContext>
    {
        TaskData taskData = new TaskData();
        
        protected override void Start()
        {
            base.Start();

            // Retain instance in navigation history until task is complete
            this.KeepAlive = true;
            
            // Launch the task
            TaskPage1 taskPage1 = new TaskPage1(this.taskData);
            taskPage1.Return += new ReturnEventHandler<TaskResult>(taskPage_Return);
            this.NavigationService.Navigate(taskPage1);
        }

        public void taskPage_Return(object sender, ReturnEventArgs<TaskResult> e)
        {
            // Task was completed (finished or canceled), return TaskResult and TaskData
            OnReturn(new ReturnEventArgs<TaskContext>(new TaskContext(e.Result, this.taskData)));
        }
    }
}
